torch_xla/experimental/stablehlo_saved_model.py

```diff
@@ -0,0 +1,126 @@
+
+  def __call__(self, *args):
+
+    call_args = []
+    for i in range(self._total_number):
+      if i in self.pos_to_orig_pos:
+        call_args.append(args[self.pos_to_orig_pos[i]])
+      else:
+        call_args.append(self.pos_to_param[i])
+
+    return tfxla.call_module(
+        tuple(call_args),
+        version=5,
+        Tout=self.tout,  # dtype information
+        Sout=self.sout,  # Shape information
+        function_list=[],
+        platforms=('CPU',),
+        module=self.model_bytecode,
+    )[0]
+
+
+def export_torch_model(model: torch.nn.Module,
+                       sample_inputs: Tuple[torch.Tensor]):
+  """Convert model into a callable backed by StableHLO.
+    
+  Args:
+      model: torch.nn.Module - a pytorch model
+      sample_inputs: Tuple[torch.Tensor] - The input to this model
+      output_shape: Tuple[int] - Shape of the output
+      output_dtype: dtype - numpy dtype for the output
+
+  This function will return a callable backed by StableHLO such that,
+
+  model(*sample_inputs) ==  export_torch_model(model, sample_inputs)(*sample_inputs)
+  (up to numerics)
+
+  In other words, returned callable have the same calling convention of the input model, and
+  on the sample input, or inputs sufficiently similar* to sample input,  
+  it is will to return same result as the original model.
+
+  * sufficiently similar input because this function will use tracing to extract the model operations
+    so it might specialize on the shapes of the sample input.
+  
+  For now, model has to only take Tensors as input and has to return 1 tensor as output.
+
+  """
+
+  # Materialize the computation before this call to make sure that StableHLO
+  # captured below only reflect the model passed in
+  xm.mark_step()
+  metrics.clear_counters()
+  model = copy.deepcopy(model)
+  sample_inputs = copy.deepcopy(sample_inputs)
+
+  device = xm.xla_device()
+  model.to(device=device)
+  sample_inputs_lazy = tuple(map(lambda x: x.to(device=device), sample_inputs))
+  input_ids = {
+      torch_xla._XLAC._xla_get_tensor_id(tensor): i
+      for i, tensor in enumerate(sample_inputs_lazy)
+  }
+  output = model(*sample_inputs_lazy)
+  if fallback_ops := dynamo_bridge.get_fallback_ops():
+    message = "This model contains ops not capturable by Pytorch/XLA: {}".format(
+        "\n".join(fallback_ops))
+    raise RuntimeError(message)
+  stablehlo_bytecode = xm.get_stablehlo_bytecode([output])
+
+  (
+      graph_input_tensor_ids,
+      graph_input_xla_values,
+  ) = torch_xla._XLAC._get_tensors_xla_device_data_node([output])
+
+  pos_to_orig_pos = {}
+  pos_to_param = {}
+  for hlo_input_pos, tensor_id in enumerate(graph_input_tensor_ids):
+    if tensor_id in input_ids:  # this is input
+      pos_to_orig_pos[hlo_input_pos] = input_ids[tensor_id]
+    else:
+      pos_to_param[hlo_input_pos] = graph_input_xla_values[
+          hlo_input_pos].detach().cpu().numpy()
+
+  output_numpy = output.cpu().detach().numpy()
```

Comment:
calling `output.cpu()` will trigger an execution(since we moved the output to cpu). Can we use `output.size` and `output.dtype` directly?